qa/L0_response_cache/ensemble_cache_test.py

```diff
@@ -0,0 +1,182 @@
+                False, "unexpected error in top-level ensemble request caching"
+            )
+
+    def test_all_models_with_cache_enabled(self):
+        self._enable_response_cache_for_all_models()
+        model_inference_stats = self._run_ensemble()
+        print(model_inference_stats)
+        if (
+            "count" not in model_inference_stats["cache_hit"]
+            or int(model_inference_stats["cache_hit"]["count"]) == 0
+        ):
+            self.assertTrue(
+                False, "unexpected error in top-level ensemble request caching"
+            )
+        if int(model_inference_stats["cache_hit"]["count"]) > int(
+            model_inference_stats["success"]["count"]
+        ):
+            self.assertTrue(False, "unexpected composing model cache hits")
+
+    def enable_cache_and_decoupled_ensemble_model(self):
+        config_file_path = os.path.join(
+            self.model_directory, self.ensemble_model_name, "config.pbtxt"
+        )
+        config_pattern = "decoupled:true"
+        config_to_add = "model_transaction_policy {\n decoupled:true\n}\n"
+        self._update_config(config_file_path, config_pattern, config_to_add)
+        config_pattern = "response_cache"
+        config_to_add = "response_cache {\n  enable:true\n}\n"
+        self._update_config(config_file_path, config_pattern, config_to_add)
+
+    def enable_composing_model_decoupled(self):
+        config_file_path = os.path.join(
+            self.model_directory, self.ensemble_model_name, "config.pbtxt"
+        )
+        config_to_remove = (
+            r"model_transaction_policy\s*\{[^}]*decoupled\s*:\s*true[^}]*\}\n*"
+        )
+        self._remove_extra_config(config_file_path, config_to_remove)
+        config_file_path = os.path.join(
+            self.model_directory, self.composing_model_name, "config.pbtxt"
+        )
+        config_pattern = "decoupled:true"
+        config_to_add = "model_transaction_policy {\n decoupled:true\n}\n"
+        self._update_config(config_file_path, config_pattern, config_to_add)
+
+    def remove_decoupled_config(self):
+        config_file_path = os.path.join(
+            self.model_directory, self.composing_model_name, "config.pbtxt"
+        )
+        config_to_remove = (
+            r"model_transaction_policy\s*\{[^}]*decoupled\s*:\s*true[^}]*\}\n*"
+        )
+        self._remove_extra_config(config_file_path, config_to_remove)
+
+    def reset_config_files(self):
+        contents = os.listdir(self.model_directory)
+        folders = [
+            folder
+            for folder in contents
+            if os.path.isdir(os.path.join(self.model_directory, folder))
+        ]
+        for model_dir in folders:
+            config_file_path = os.path.join(
+                self.model_directory, str(model_dir), "config.pbtxt"
+            )
+            config_to_remove = r"response_cache\s*\{[^}]*\}\n*"
+            self._remove_extra_config(config_file_path, config_to_remove)
```

Comment:
Add leading `_` to helper functions that aren't starting with `test_` for consistency.

Reply:
Will fix this.